base/experiments/ad9910_ram.py

```diff
@@ -0,0 +1,206 @@
+        match ram_type:
+            case RAMType.FREQ:
+                self.dest = RAM_DEST_FTW
+                dds.frequency_to_ram(data, ram)
+                self.asf = dds.amplitude_to_asf(base_amplitude)
+                self.pow = dds.turns_to_pow(base_phase)
+            case RAMType.PHASE:
+                self.dest = RAM_DEST_POW
+                dds.turns_to_ram(data, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+                self.asf = dds.amplitude_to_asf(base_amplitude)
+            case RAMType.AMP:
+                self.dest = RAM_DEST_ASF
+                dds.amplitude_to_ram(data, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+                self.pow = dds.turns_to_pow(base_phase)
+            case RAMType.POLAR:
+                self.dest = RAM_DEST_POWASF
+                phase, amp = data
+                dds.turns_amplitude_to_ram(phase, amp, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+            case _:
+                raise NotImplementedError
+
+        self.start_addr = 0
+        self.end_addr = len(ram) - 1    # Inclusive
+
+        # Configure ther use of OSK. Note that using different OSK settings for
+        # different channels introduces step swiching latencies
+        if ram_type == RAMType.AMP or ram_type == RAMType.POLAR:
+            # Disable OSK. OSK amplitude as it has a higher priority than RAM.
+            self.osk_enable = 0
+        else:
+            # Enable OSK. There are no other utilizable amplitude data source.
+            self.osk_enable = 1
+
+        # Conversion to list to avoid type mismatch in the kernel
+        self.ram = ram.tolist()
+
+        # Note: Integer conversion may cause inaccuracy
+        self.step = int(ramp_interval * dds.sysclk / 4.0)
+
+        self.ramp_mode = ramp_mode
+
+
+class RAMProfileMap:
+    def __init__(self, core):
+        """Initialize a RAM profile to DDS mapping
+
+        Args:
+            core: The ARTIQ Core instance for time control.
+        """
+        self.ram_profile_map = []
+        self.cplds = []
+        self.core = core
+
+    def append(self, dds, ram_profile):
+        """Append a RAM profile into the builder. In addition, register the
+            CPLDs that have DDSes that playback the RAM profile.
+
+        Args:
+            dds: The DDS that will playback the RAM profile.
+            ram_profile: The RAM profile.
+        """
+        self.ram_profile_map.append((dds, ram_profile))
+
+        # Add the associlated CPLD to the list if not already there.
+        # ARTIQ-Python does not support iterration of a Python builtin set
+        # The workaround is to maintain a list, but convert it into set to
+        # avoid CPLD duplications.
+        cplds = set(self.cplds)
+        cplds.add(dds.cpld)
+        self.cplds = list(cplds)
+
+    @kernel
+    def program(self):
+        """Program the RAM profiles to the corresponding AD9910s.
```

Comment:
I still do not fully follow the code here. At what line is the DDS on? When does it start/stop RAM modulation? Does the code ever set the RAM back to the single tone mode?

Reply:
https://github.com/Jayich-Lab/jax/blob/bf89fc1c21ccdc87dc309a458175fd630bb75b9e/base/experiments/ad9910_ram.py#L187-L188

All RAM profiles are activated after the I/O updates.
Though there is probably a slight bug where I/O updates of different Urukuls are triggered sequentially instead of in parallel. The `delay()` call in `pulse_mu()` does advance the timeline.